privtrak.go

```diff
@@ -281,14 +311,70 @@ func New(provided Config, identifier UserIdentifier, handler DeltaHandler) (midd
 		}
 	}()
 
+	// Start a goroutine for flushing seedtimes
+	mw.wg.Add(1)
+	go func() {
+		defer mw.wg.Done()
+		t := time.NewTicker(cfg.SeedtimeFlushInterval)
+		for {
+			select {
+			case <-mw.closing:
+				t.Stop()
+				return
+			case <-t.C:
+				before := time.Now()
+				mw.flushAllSeedTimes()
+				log.Debug("privtrak: populateProm() finished", log.Fields{"timeTaken": time.Since(before)})
+			}
+		}
+	}()
+
 	return &mw, nil
 }
 
 func (m *ptMiddleware) collectGarbage() {
 	for _, shard := range m.shards {
 		shard.Lock()
 		for id, stats := range shard.users {
-			stats.removeExpired(m.cfg.PeerLifetime)
+
+			expiredDeltas := stats.removeExpired(m.cfg.PeerLifetime, id)
+			shard.deltas = append(shard.deltas, expiredDeltas...)
+
+			if len(stats.swarmStats) == 0 {
+				delete(shard.users, id)
+			} else {
+				shard.users[id] = stats
+			}
+		}
+		shard.Unlock()
+	}
+}
+
+func (u *userStats) flushSeedTimes(id ID) []StatDelta {
+	userDeltas := make([]StatDelta, len(u.swarmStats))
+
+	for i, stats := range u.swarmStats {
+		delta := StatDelta{
+			InfoHash:    stats.infoHash,
+			Reported:    timecache.Now(),
+			SeedingTime: stats.seedingTime,
+			User: id,
+		}
+		u.swarmStats[i].seedingTime = 0
+		userDeltas = append(userDeltas, delta)
+	}
+	return userDeltas
+}
+
+func (m *ptMiddleware) flushAllSeedTimes() {
+
+	for _, shard := range m.shards {
+		shard.Lock()
+		for id, stats := range shard.users {
+
+			partialDeltas := stats.flushSeedTimes(id)
+			shard.deltas = append(shard.deltas, partialDeltas...)
+
 			if len(stats.swarmStats) == 0 {
```

Comment:
This check is not necessary here because `flushSeedTimes` does not delete stats.